cvmfs/cache_ram.h

```diff
@@ -0,0 +1,191 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+#ifndef CVMFS_CACHE_RAM_H_
+#define CVMFS_CACHE_RAM_H_
+
+#include <pthread.h>
+
+#include <cassert>
+#include <cstdlib>
+#include <vector>
+
+#include "statistics.h"
+#include "cache.h"
+#include "hash.h"
+#include "util/pointer.h"
+#include "kvstore.h"
+
+namespace cache {
+
+/**
+ * ...
+ * TODO(jblomer): save open file table for hotpatch
+ */
+class RamCacheManager : public CacheManager {
+ public:
+  virtual CacheManagerIds id() { return kRamCacheManager; }
+
+  RamCacheManager(uint64_t max_size, unsigned max_entries, perf::Statistics *statistics)
+    : max_size(max_size)
+    , invalid_fd_(shash::Any())
+    , pinned_entries_(max_entries/3, "RamCache.pinned", statistics)
+    , regular_entries_(max_entries/3, "RamCache.regular", statistics)
+    , volatile_entries_(max_entries/3, "RamCache.volatile", statistics) {
+    int retval = pthread_rwlock_init(&rwlock_, NULL);
+    assert(retval == 0);
+  }
+  virtual ~RamCacheManager() {
+    pthread_rwlock_destroy(&rwlock_);
+  }
+  virtual bool AcquireQuotaManager(QuotaManager *quota_mgr);
+
+  /**
+   * Open a new file descriptor into the cache. Note that opening entries effectively pins them in the cache,
+   * so it may be necessary to close unneeded file descriptors to allow eviction to make room in the cache
+   * @param id The hash key
+   * @returns The smallest free integer file descriptor, or -ENOENT if the entry
+   * is not in the cache
+   */
+  virtual int Open(const shash::Any &id);
+
+  /**
+   * Look up the size in bytes of the open cache entry
+   * @param id The hash key
+   * @returns The size of the entry, or -EBADFD if fd is not valid
+   */
+  virtual int64_t GetSize(int fd);
+
+  /**
+   * Close the descriptor in the cache. Entries not marked as pinned will become subject to
+   * eviction once closed
+   * @param id The hash key
+   * @returns -EBADFD if fd is not valid
+   */
+  virtual int Close(int fd);
+
+  /**
+   * Read a section from the cache entry. See pread(3) for a discussion of the arguments
+   * @param id The hash key
+   * @returns The number of bytes copied, or -EBADFD if fd is not valid
+   */
+  virtual int64_t Pread(int fd, void *buf, uint64_t size, uint64_t offset);
+
+  /**
+   * Duplicates the open file descriptor, allowing the original and the new one to be
+   * used independently
+   * @param id The hash key
+   * @returns A new fd, or -EBADFD if fd is not valid
+   */
+  virtual int Dup(int fd);
+
+  /**
+   * No effect for in-memory caches
+   * @param id The hash key
+   * @returns -EBADFD if fd is not valid
+   */
+  virtual int Readahead(int fd);
+
+  /** Get the amount of space to be allocated for a call to @ref StartTxn */
```

Comment:
Is this still Doxygen syntax?  I'm not sure, if it is, it's good to know that one can do one-liners this way.